From 33909d72d2f31f18b7106d3af8d8c8bfc37a5024 Mon Sep 17 00:00:00 2001
From: Tobias <ruck.tobias@gmail.com>
Date: Thu, 22 Aug 2024 19:21:13 +0200
Subject: [PATCH] Add `BlockManager::ReadBlockHeaderFromDisk` (#43)

This is needed for auxpow-aware `CBlockIndex::GetBlockHeader`.

(cherry picked from commit 6d6ec03d9f27b82c03cf51e6ef09584861428ee7)

diff --git a/src/node/blockstorage.cpp b/src/node/blockstorage.cpp
index 582f8f113b113b3837d84a9a6bab7de4573bb888..839ed38b034f1c9928976b06115753b9a390da58 100644
--- a/src/node/blockstorage.cpp
+++ b/src/node/blockstorage.cpp
@@ -1029,6 +1029,51 @@ bool BlockManager::ReadRawBlock(std::vector<uint8_t> &block,
     return true;
 }
 
+bool BlockManager::ReadBlockHeaderFromDisk(CBlockHeader &header,
+                                           const FlatFilePos &pos) const {
+    header.SetNull();
+
+    // Open history file to read
+    CAutoFile filein(OpenBlockFile(pos, true), SER_DISK, CLIENT_VERSION);
+    if (filein.IsNull()) {
+        return error("ReadBlockHeaderFromDisk: OpenBlockFile failed for %s",
+                     pos.ToString());
+    }
+
+    // Read header
+    try {
+        filein >> header;
+    } catch (const std::exception &e) {
+        return error("%s: Deserialize or I/O error - %s at %s", __func__,
+                     e.what(), pos.ToString());
+    }
+
+    // Check the header
+    if (!CheckAuxProofOfWork(header, GetConsensus())) {
+        return error("ReadBlockHeaderFromDisk: Errors in block header at %s",
+                     pos.ToString());
+    }
+
+    return true;
+}
+
+bool BlockManager::ReadBlockHeaderFromDisk(CBlockHeader &header,
+                                           const CBlockIndex &index) const {
+    const FlatFilePos block_pos{WITH_LOCK(cs_main, return index.GetBlockPos())};
+
+    if (!ReadBlockHeaderFromDisk(header, block_pos)) {
+        return false;
+    }
+
+    if (header.GetHash() != index.GetBlockHash()) {
+        return error("ReadBlockHeaderFromDisk(CBlock&, CBlockIndex*): "
+                     "GetHash() doesn't match index for %s at %s",
+                     index.ToString(), block_pos.ToString());
+    }
+
+    return true;
+}
+
 bool BlockManager::ReadTxFromDisk(CMutableTransaction &tx,
                                   const FlatFilePos &pos) const {
     // Open history file to read
diff --git a/src/node/blockstorage.h b/src/node/blockstorage.h
index 0f769e2dd0698f80b8309e328c49876cc04d130c..5f0f2069e3195f88c18dd8ec113846095dc22f07 100644
--- a/src/node/blockstorage.h
+++ b/src/node/blockstorage.h
@@ -437,6 +437,9 @@ public:
      */
     void UnlinkPrunedFiles(const std::set<int> &setFilesToPrune) const;
 
+    bool ReadBlockHeaderFromDisk(CBlockHeader &header, const FlatFilePos &pos) const;
+    bool ReadBlockHeaderFromDisk(CBlockHeader &header, const CBlockIndex &index) const;
+
     /** Functions for disk access for blocks */
     bool ReadBlock(CBlock &block, const FlatFilePos &pos) const;
     bool ReadBlock(CBlock &block, const CBlockIndex &index) const;
diff --git a/src/test/blockstorage_tests.cpp b/src/test/blockstorage_tests.cpp
index 5edef5612e03d702a5a5ad3264a05d616eae735a..4df8dcfe11ccb1f2b2fb20ab6bee8bd9b0641c1e 100644
--- a/src/test/blockstorage_tests.cpp
+++ b/src/test/blockstorage_tests.cpp
@@ -48,6 +48,8 @@ BOOST_AUTO_TEST_CASE(read_tx_data_from_disk) {
         CBlockIndex *pindex = active_tip->GetAncestor(height);
         CBlock block;
         BOOST_CHECK(chainman.m_blockman.ReadBlock(block, *pindex));
+        CBlockHeader header;
+        BOOST_CHECK(chainman.m_blockman.ReadBlockHeaderFromDisk(header, *pindex));
         // + 81 = CBlockHeader + CompactSize
         CheckReadTx(
             chainman.m_blockman,
